neqo-http3-server/src/old_https.rs

```diff
@@ -0,0 +1,180 @@
+}
+
+impl Http09Server {
+    pub fn new(
+        now: Instant,
+        certs: &[impl AsRef<str>],
+        protocols: &[impl AsRef<str>],
+        anti_replay: AntiReplay,
+        cid_manager: Rc<RefCell<dyn ConnectionIdManager>>,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            server: Server::new(
+                now,
+                certs,
+                protocols,
+                anti_replay,
+                Box::new(AcceptZeroRttChecker {}),
+                cid_manager,
+            )?,
+            conn_state: HashMap::new(),
+        })
+    }
+
+    fn stream_readable(&mut self, stream_id: u64, mut conn: &mut ActiveConnectionRef, args: &Args) {
+        if stream_id % 4 != 0 {
+            eprintln!("Stream {} not client-initiated bidi, ignoring", stream_id);
+            return;
+        }
+        let mut data = vec![0; 4000];
+        conn.borrow_mut()
+            .stream_recv(stream_id, &mut data)
+            .expect("Read should succeed");
+        let msg = match String::from_utf8(data) {
+            Ok(s) => s,
+            Err(_e) => {
+                eprintln!("invalid string. Is this HTTP 0.9?");
+                conn.borrow_mut().stream_close_send(stream_id).unwrap();
+                return;
+            }
+        };
+        let re = if args.qns_mode {
+            Regex::new(r"GET +/(\S+)(\r)?\n").unwrap()
+        } else {
+            Regex::new(r"GET +/(\d+)(\r)?\n").unwrap()
+        };
+        let m = re.captures(&msg);
+        let resp = match m.and_then(|m| m.get(1)) {
+            None => Some(b"Hello World".to_vec()),
+            Some(path) => {
+                let path = path.as_str();
+                eprintln!("Path = '{}'", path);
+                if args.qns_mode {
+                    qns_read_response(path)
+                } else {
+                    let count = usize::from_str_radix(path, 10).unwrap();
+                    Some(vec![b'a'; count])
+                }
+            }
+        };
+        let conn_state = self.conn_state.get_mut(&(conn.clone(), stream_id)).unwrap();
+        conn_state.data_to_send = resp.map(|r| (r, 0));
+        if conn_state.writable {
+            self.stream_writable(stream_id, &mut conn);
+        }
+    }
+
+    fn stream_writable(&mut self, stream_id: u64, conn: &mut ActiveConnectionRef) {
+        match self.conn_state.get_mut(&(conn.clone(), stream_id)) {
+            None => {
+                eprintln!("Unknown stream {}, ignoring event", stream_id);
+            }
+            Some(conn_state) => {
+                conn_state.writable = true;
+                if let Some((data, mut offset)) = &mut conn_state.data_to_send {
+                    let sent = conn
+                        .borrow_mut()
+                        .stream_send(stream_id, &data[offset..])
+                        .unwrap();
+                    eprintln!("Wrote {}", sent);
+                    offset += sent;
+                    if offset == data.len() {
```

Comment:
I see why you aren't using `Vec::split_off` for this, but we could use a slice instead and avoid the need to use a tuple.  Then you could do
```rust
if let Some(ref mut remaining) = &mut conn_state.data_to_send {
    let sent = ...;
    remaining = remaining[sent..];
    if remaining.is_empty() {
        conn.borrow_mut().stream_close_send()
    } etc...

Reply:
A slice is a reference, and we need to own the data in `conn_state`.